src/upgrade.py

```diff
@@ -0,0 +1,135 @@
+        """
+        return self.cluster_state == "idle"
+
+    @cached_property
+    def client(self) -> ZooKeeperManager:
+        """Cached client manager application for performing ZK commands."""
+        return ZooKeeperManager(
+            hosts=self.charm.cluster.active_hosts,
+            client_port=self.charm.cluster.client_port,
+            username="super",
+            password=self.charm.cluster.passwords[0],
+        )
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        default_message = "Pre-upgrade check failed and cannot safely upgrade"
+        try:
+            if not self.client.members_broadcasting or not len(self.client.server_members) == len(
+                self.charm.cluster.peer_units
+            ):
+                raise ClusterNotReadyError(
+                    message=default_message,
+                    cause="Not all application units are connected and broadcasting in the quorum",
+                )
+
+            if self.client.members_syncing:
+                raise ClusterNotReadyError(
+                    message=default_message, cause="Some quorum members are syncing data"
+                )
+
+            if not self.charm.cluster.stable:
+                raise ClusterNotReadyError(
+                    message=default_message, cause="Charm has not finished initialising"
+                )
+
+        except QuorumLeaderNotFoundError:
+            raise ClusterNotReadyError(message=default_message, cause="Quorum leader not found")
+        except ConnectionClosedError:
+            raise ClusterNotReadyError(
+                message=default_message, cause="Unable to connect to the cluster"
+            )
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        upgrade_stack = []
+        for unit in self.charm.cluster.peer_units:
+            config = self.charm.cluster.unit_config(unit=unit)
+
+            # upgrade quorum leader last
+            if config["host"] == self.client.leader:
+                upgrade_stack.insert(0, int(config["unit_id"]))
+            else:
+                upgrade_stack.append(int(config["unit_id"]))
+
+        return upgrade_stack
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        logger.warning("SOME USEFUL INSTRUCTIONS")  # TODO: figure it out
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        self.charm.snap.stop_snap_service()
+
+        if not self.charm.snap.install():
+            logger.error("Unable to install ZooKeeper Snap")
+            self.set_unit_failed()
+            return
+
+        logger.info(f"{self.charm.unit.name} upgrading service...")
+        self.charm.snap.restart_snap_service()
+
+        try:
+            logger.debug("Running post-upgrade check...")
+            self.pre_upgrade_check()
+
+            logger.debug("Marking unit completed...")
+            self.set_unit_completed()
```

Comment:
**thought** could we abstract this to the library instead? I mean, I suppose these lines of code (also the 129-131) are something that every charm should do. "If the UpgradeGrantedEvent runs successfully set_unit_completed, otherwise log the error and set_unit_failed" seems something very general. Can there be any way we can factor this out such that they are implicit in the library?

Reply:
I agree with your point entirely, there's probably a lot of common patterns in the VM + K8s implementation that can be abstracted away from the authors.

IMO right now it's hard to know when that abstraction is useful, as for now it's still only really been worked on by a few people. I think after more products have used the lib, it's a better time to see what can be brought in. 

Avoids early-optimisation, and hopefully abstraction pain.